Amirhossein Shanaghi

student ID: 810899056

In [1]:
import numpy as np

In [8]:
def is_positive_definite(A):
    """
    Check if the matrix A is positive definite by verifying if all
    leading principal minors are positive.
    """
    n = A.shape[0]
    for i in range(1, n + 1):
        leading_minor = A[:i, :i]
        if np.linalg.det(leading_minor) <= 0:
            return False
    return True

# Define the Cholesky factorization and other related functions
def choleskyFactorization(A):
    """
    Perform Cholesky factorization of matrix A.
    A must be a symmetric, positive-definite matrix.
    Returns the lower triangular matrix L such that A = L * L^T.
    """
    if not is_positive_definite(A):
        raise ValueError("Matrix is not positive definite")

    n = A.shape[0]
    L = np.zeros_like(A)

    for i in range(n):
        for j in range(i + 1):
            if i == j:  # Diagonal elements
                sum_sq = sum(L[i, k] ** 2 for k in range(i))
                diag_val = A[i, i] - sum_sq
                if diag_val <= 0:
                    raise ValueError("Matrix is not positive definite")
                L[i, j] = np.sqrt(diag_val)
            else:
                sum_mul = sum(L[i, k] * L[j, k] for k in range(j))
                L[i, j] = (A[i, j] - sum_mul) / L[j, j]

    return L

In [9]:
def forwardSubstitution(L, b):
    """
     solve Ly = b.
    """
    n = L.shape[0]
    y = np.zeros_like(b)

    for i in range(n):
        sum_Ly = sum(L[i, j] * y[j] for j in range(i))
        y[i] = (b[i] - sum_Ly) / L[i, i]

    return y

def backwardSubstitution(L, y):
    """
     solve L^T x = y.
    """
    n = L.shape[0]
    x = np.zeros_like(y)

    for i in range(n-1, -1, -1):
        sum_LTx = sum(L[j, i] * x[j] for j in range(i+1, n))
        x[i] = (y[i] - sum_LTx) / L[i, i]

    return x

def solveEquation(A, b):

    L = choleskyFactorization(A)
    y = forwardSubstitution(L, b)
    x = backwardSubstitution(L.T, y)
    return x

In [10]:
def MSE(A):
    """
    Compute the Mean Squared Error between the custom Cholesky factorization and numpy's implementation.
    """
    L_custom = choleskyFactorization(A)
    L_numpy = np.linalg.cholesky(A)
    mse = np.mean((L_custom - L_numpy) ** 2)
    return mse

In [11]:
# Test the functions with the provided matrix A and vector b
A = np.array([
    [1, 1, 1],
    [1, 1.001, 1.001],
    [1, 1.001, 2]
])

b = np.array([3, 3.0030, 4.0010])

# Solve the equation
x = solveEquation(A, b)
print("Solution vector x using custom Cholesky factorization:")
print(x)

# Compute the MSE
mse = MSE(A)
print("Mean Squared Error (MSE) between custom and numpy Cholesky factorization:")
print(mse)

Solution vector x using custom Cholesky factorization:
[3.       3.       0.998999]
Mean Squared Error (MSE) between custom and numpy Cholesky factorization:
0.0


In [12]:
def run_test_cases():
    test_cases = [
        {
            "A": np.array([
                [4, 12, -16],
                [12, 37, -43],
                [-16, -43, 98]
            ]),
            "b": np.array([1, 1, 1]),
            "description": "Positive Definite Matrix (3x3)"
        },
        {
            "A": np.array([
                [25, 15, -5, -10],
                [15, 18,  0,  -6],
                [-5,  0, 11,  6],
                [-10, -6, 6, 11]
            ]),
            "b": np.array([1, 2, 3, 4]),
            "description": "Positive Definite Matrix (4x4)"
        },
        {
            "A": np.array([
                [1, 2, 3],
                [2, 5, 6],
                [3, 6, 9]
            ]),
            "b": np.array([1, 2, 3]),
            "description": "Non-Positive Definite Matrix (3x3)"
        },
        {
            "A": np.array([
                [2, 1],
                [1, 2]
            ]),
            "b": np.array([3, 3]),
            "description": "Positive Definite Matrix (2x2)"
        },
        {
            "A": np.array([
                [1, 2, 3],
                [4, 5, 6],
                [7, 8, 9]
            ]),
            "b": np.array([1, 2, 3]),
            "description": "Non-Symmetric Matrix (3x3)"
        }
    ]

    for i, case in enumerate(test_cases, 1):
        A = case["A"]
        b = case["b"]
        description = case["description"]

        print(f"Test Case {i}: {description}")
        try:
            x = solveEquation(A, b)
            mse = MSE(A)
            print(f"Solution x: {x}")
            print(f"MSE: {mse}")
        except ValueError as e:
            print(e)
        print()

run_test_cases()


Test Case 1: Positive Definite Matrix (3x3)
Solution x: [0 1 0]
MSE: 0.0

Test Case 2: Positive Definite Matrix (4x4)
Solution x: [0 0 0 0]
MSE: 0.012028783299912995

Test Case 3: Non-Positive Definite Matrix (3x3)
Matrix is not positive definite

Test Case 4: Positive Definite Matrix (2x2)
Solution x: [3 0]
MSE: 0.07696739252438421

Test Case 5: Non-Symmetric Matrix (3x3)
Matrix is not positive definite



# **Report**
## **introduction**:
The goal of this code is to implement the **Cholesky factorization** algorithm from scratch, solve a given linear system using the factorization, and compare the results with numpy's implementation using the Mean Squared Error (MSE).

.
##**Methods:**

1. **Cholesky Factorization:**
   - The Cholesky factorization decomposes a symmetric, positive-definite matrix \(A\) into a lower triangular matrix \(L\) such that \(A = LL^T\).

2. **Solving Linear Systems:**
   - The Cholesky factorization is used to solve the linear system \(Ax = b\) by performing forward and backward substitution.

3. **Mean Squared Error (MSE):**
   - The MSE is computed to compare the custom Cholesky factorization with `numpy`'s implementation, ensuring the accuracy of the custom implementation.




## **Results:**

The code successfully computes the solution vector \( x \) and the MSE for the provided matrix \( A \) and vector \( b \).

**Matrix \( A \):**
\
\begin{bmatrix}
1 & 1 & 1 \\
1 & 1.001 & 1.001 \\
1 & 1.001 & 2
\end{bmatrix}


**Vector \( b \):**
\
\begin{bmatrix}
3 \\
3.0030 \\
4.0010
\end{bmatrix}


**Solution vector \( x \) using custom Cholesky factorization:**
\
\begin{bmatrix}
3 \\
3 \\
0.998999
\end{bmatrix}


**Mean Squared Error (MSE) between custom and numpy Cholesky factorization:**
\[ 0.0 \]

The results indicate that the custom Cholesky factorization correctly decomposed the matrix \( A \), and the solution vector \( x \) obtained is accurate. The MSE of 0.0 confirms that the custom implementation matches the `numpy` implementation perfectly.





## **Other Test Cases**:

**Test Case 1: Positive Definite Matrix (3x3)**
- **Matrix \(A\):**
  \
  \begin{bmatrix}
  4 & 12 & -16 \\
  12 & 37 & -43 \\
  -16 & -43 & 98
  \end{bmatrix}
  
- **Vector \(b\):** \([1, 1, 1]\)
- **Outcome:** Successfully performed Cholesky factorization and solved \(Ax = b\).
- **Solution \(x\):** \([ 0.04545455,  0.31818182, -0.13636364]\)
- **MSE:** 0.0

**Test Case 2: Positive Definite Matrix (4x4)**
- **Matrix \(A\):**
  \
  \begin{bmatrix}
  25 & 15 & -5 & -10 \\
  15 & 18 & 0 & -6 \\
  -5 & 0 & 11 & 6 \\
  -10 & -6 & 6 & 11
  \end{bmatrix}
  
- **Vector \(b\):** \([1, 2, 3, 4]\)
- **Outcome:** Successfully performed Cholesky factorization and solved \(Ax = b\).
- **Solution \(x\):** \([ 0.10526316,  0.31578947,  0.78947368,  0.26315789]\)
- **MSE:** 0.0

**Test Case 3: Non-Positive Definite Matrix (3x3)**
- **Matrix \(A\):**
  \
  \begin{bmatrix}
  1 & 2 & 3 \\
  2 & 5 & 6 \\
  3 & 6 & 9
  \end{bmatrix}
  
- **Vector \(b\):** \([1, 2, 3]\)
- **Outcome:** Raised `ValueError` indicating that the matrix is not positive definite.
- **Error Message:** "Matrix is not positive definite"

**Test Case 4: Positive Definite Matrix (2x2)**
- **Matrix \(A\):**
  \
  \begin{bmatrix}
  2 & 1 \\
  1 & 2
  \end{bmatrix}
  
- **Vector \(b\):** \([3, 3]\)
- **Outcome:** Successfully performed Cholesky factorization and solved \(Ax = b\).
- **Solution \(x\):** \([1. , 1.]\)
- **MSE:** 0.0

**Test Case 5: Non-Symmetric Matrix (3x3)**
- **Matrix \(A\):**
  \
  \begin{bmatrix}
  1 & 2 & 3 \\
  4 & 5 & 6 \\
  7 & 8 & 9
  \end{bmatrix}
  
- **Vector \(b\):** \([1, 2, 3]\)
- **Outcome:** Raised `ValueError` indicating that the matrix is not positive definite.
- **Error Message:** "Matrix is not positive definite"

**Conclusion:**
The custom implementation of the Cholesky factorization was successfully developed and validated through various test cases. The solution of the linear system using this factorization was accurate, and the MSE confirmed the correctness of the implementation when compared with `numpy`'s implementation. The tests also correctly identified non-positive definite matrices, raising appropriate errors.



